### DISCUSSION TITLE
Traverse Array of Aff to Aff of array

### DISCUSSION THREAD

**Discussion by azhdari** [louthy/language-ext#994] [Q&A] [answered]
Hi,

How can we implement this function?

`static Aff<RT, Arr<A>> TraverseArray<RT, A>(this Arr<Aff<RT, A>> arr)`

Maybe it's possible using `LanguageExt.Aff<R, Arr<A>>.Effect();`, but I don't know how.

Thank you.

**Comment by CK-LinoPro** [comment]
`Aff` is an asynchronous monad while `Arr` is a synchronous monad.
It is usually not possible to implement `sequence :: t (m a) -> m (t a)` (the function you want under its usual name) for this case.

Let's consider the simple case of `static Option<Task<T>> Sequence(this Task<Option<T>> task)` instead.
`Task` is asynchronous, `Option` is synchronous.

``` cs
static Option<Task<T>> Sequence<T>(this Task<Option<T>> task)
    => ??? : Some(???) : None;
```

The challenge is filling in the blanks (`???`) but it turns out that's impossible because to know whether the task holds a `Some` or a `None` value you have to _wait for the task to finish_ for which you need an asynchronous result type.

The other way around (synchronous -> asynchronous) is trivial.

``` cs
static async Task<Option<T>> Sequence(this Option<Task<T>> option)
    => option.Case switch
    {
        T t => Some((await t).ConfigureAwait(false)),
        _ => None,
    };
```

**Reply by azhdari** [reply]
Thank you for your time. What you mentioned is exactly what I was thinking about. I'm so glad that it has a build-in solution.

**Comment by louthy** [comment, marked as answer]
You can use the built-in methods:

* SequenceSerial
* SequenceParallel
* TraverseSerial
* TraverseParallel

**Reply by azhdari** [reply]
WOW, thank you so much. I don't know how I didn't see these methods!
